### ISSUE TITLE
Add pgtype.Uint64 to support pg_lsn type

### ISSUE THREAD

**Issue by hnakamur** [jackc/pgx#1407]
I'd like to add `pgtype.Uint64` to support `pg_lsn` type.

I read https://github.com/jackc/pgx/issues/462#issuecomment-526838739 and noticed type LSN is defined at https://github.com/jackc/pglogrepl/blob/1e972958ba0339bc189fe5d5b190ce97f0258e3f/pglogrepl.go#L48.

However I think it's more consistent to add `pgtype.Uint64` since `pgtype.Uint32` was added to replace CID, OID, OIDValue, and XID at
https://github.com/jackc/pgx/commit/eec82c9433368eaf9855ad6ee0a4f36418c6318a

My attempt to add pgtype.Uint64 to support pg_lsn type is at https://github.com/jackc/pgx/commit/e925c2eebcd02dd16ad702da194c79c85000b51b

With this change, I can define `PgLSN` type like below just as described at https://pkg.go.dev/github.com/jackc/pgx/v5@v5.1.1/pgtype#hdr-Extending_Existing_Type_Support

```
type PgLSN struct {
	LSN   uint64
	Valid bool
}

func (n *PgLSN) ScanUint64(v pgtype.Uint64) error {
	*n = PgLSN{
		LSN:   v.Uint64,
		Valid: v.Valid,
	}
	return nil
}

func (n PgLSN) Uint64Value() (pgtype.Uint64, error) {
	return pgtype.Uint64{
		Uint64: n.LSN,
		Valid:  true,
	}, nil
}

func (src *PgLSN) String() string {
	if !src.Valid {
		return ""
	}
	return fmt.Sprintf("%X/%X", src.LSN>>32, uint32(src.LSN))
}

type PgLSNCodec struct {
	pgtype.Uint64Codec
}

func (PgLSNCodec) DecodeValue(tm *pgtype.Map, oid uint32, format int16, src []byte) (interface{}, error) {
	if src == nil {
		return nil, nil
	}

	var target PgLSN
	scanPlan := tm.PlanScan(oid, format, &target)
	if scanPlan == nil {
		return nil, fmt.Errorf("PlanScan did not find a plan")
	}

	err := scanPlan.Scan(src, &target)
	if err != nil {
		return nil, err
	}

	return target, nil
}
```

**Comment by jackc** [comment]
Are there any other types in PostgreSQL that would benefit from this generic pgtype.Uint64 type like pgtype.Uint32 was used by CID, OID, OIDValue, and XID? 

Also, would the text format be compatible? `16/B374D848` doesn't look like a normal number. If there are other uint64 types they may be formatted differently.

**Comment by hnakamur** [comment]
Thanks for taking time to response.

> Are there any other types in PostgreSQL that would benefit from this generic pgtype.Uint64 type like pgtype.Uint32 was used by CID, OID, OIDValue, and XID?

Such another type is [FullTransactionId](https://github.com/postgres/postgres/blob/66456da1505b8347a6d2ddd23b34e62e904c4908/src/include/access/transam.h#L65-L68).

I found it with running `grep -r 'return .*DatumGetUInt64('` in the PostgreSQL source code.
Surprisingly `DatumGetLSN` is implemented with `DatumGetInt64` instead of `DatumGetUInt64`
https://github.com/postgres/postgres/blob/66456da1505b8347a6d2ddd23b34e62e904c4908/src/include/utils/pg_lsn.h#L24
even though `XLogRecPtr` is typedefed to `uint64`.
https://github.com/postgres/postgres/blob/66456da1505b8347a6d2ddd23b34e62e904c4908/src/include/access/xlogdefs.h#L21

Since `Int8` is validated if the value is between `math.MinInt8` and `math.MaxInt8`,
https://github.com/jackc/pgx/blob/88b373f9eed6b0cecb8b12d95999578bdfdbc533/pgtype/int.go#L1443-L1447
 I suppose we need pgtype.Uint64 to be added. Is that right?

> Also, would the text format be compatible? 16/B374D848 doesn't look like a normal number. If there are other uint64 types they may be formatted differently.

I suppose the text format of `pgtype.Uint64` should be the plain decimal. And we can define `PgLSN` as an extended type of `pgtype.Uint64` and override the text format to be like `16/B374D848`.

**Comment by hnakamur** [comment]
Thanks!
Is it OK to add `pgtype.PgLSN` too?
I think for most case we would like to use it instead of plain `pgtype.Uint64`.

I made a commit to do so: https://github.com/jackc/pgx/commit/aa8bbb4b2ed44489fa32d0f3003c12662e25f950.
I confirmed tests for both `Uint64Codec` and `PgLSN` passes as well as all existing tests.

```
$ go test -v ./... | grep -E '(Uint64Codec|PgLSN)'
=== RUN   TestConnQuerPgLSN
=== PAUSE TestConnQuerPgLSN
=== CONT  TestConnQuerPgLSN
--- PASS: TestConnQuerPgLSN (0.01s)
=== RUN   TestUint64Codec
--- PASS: TestUint64Codec (0.00s)
$ go version
go version go1.19.3 linux/amd64
$ dpkg-query --show postgresql-15
postgresql-15   15.1-1.pgdg22.04+1
$ grep ^PRETTY_NAME= /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
```

**Comment by hnakamur** [comment]
I opened a pull request: [Add pgtype.Uint64 to support pgtype.PgLSN type (closes #1407) by hnakamur · Pull Request #1416 · jackc/pgx](https://github.com/jackc/pgx/pull/1416).
Could you review that in your spare time?
Thanks.

**Comment by hnakamur** [comment]
Closing this after https://github.com/jackc/pgx/pull/1416#discussion_r1045104071
